packages/babel-plugin-transform-react-display-name/src/index.js

```diff
@@ -2,28 +2,116 @@ import { declare } from "@babel/helper-plugin-utils";
-      const prop = props[i];
-      const key = t.toComputedKey(prop);
-      if (t.isLiteral(key, { value: "displayName" })) {
-        safe = false;
-        break;
-      }
+function addDisplayNameInCreateClass(id, call) {
+  const props = call.arguments[0].properties;
+  let safe = true;
+
+  for (let i = 0; i < props.length; i++) {
+    const prop = props[i];
+    const key = t.toComputedKey(prop);
+    if (t.isLiteral(key, { value: "displayName" })) {
+      safe = false;
+      break;
     }
+  }
 
-    if (safe) {
-      props.unshift(
-        t.objectProperty(t.identifier("displayName"), t.stringLiteral(id)),
-      );
+  if (safe) {
+    props.unshift(
+      t.objectProperty(t.identifier("displayName"), t.stringLiteral(id)),
+    );
+  }
+}
+
+function getDisplayNameReferenceIdentifier(
+  path: NodePath<t.CallExpression>,
+): ?t.Identifier {
+  let id;
+
+  // crawl up the ancestry looking for possible candidates for displayName inference
+  path.find(function (path) {
+    if (path.isAssignmentExpression()) {
+      id = path.node.left;
+    } else if (path.isObjectProperty()) {
+      id = path.node.key;
+    } else if (path.isVariableDeclarator()) {
+      id = path.node.id;
+    } else if (path.isStatement()) {
+      // we've hit a statement, we should stop crawling up
+      return true;
     }
+
+    // we've got an id! no need to continue
+    if (id) return true;
+  });
+
+  // ensure that we have an identifier we can inherit from
+  if (!id) return;
+
+  // foo.bar -> bar
+  if (t.isMemberExpression(id)) {
+    id = id.property;
+  }
+
+  // identifiers are the only thing we can reliably get a name from
+  if (!t.isIdentifier(id)) return;
+
+  return id;
+}
+
+function isCreateContext(node) {
+  let callee;
+  return (
+    t.isCallExpression(node) &&
+    t.isMemberExpression((callee = node.callee)) &&
+    t.isIdentifier(callee.object, { name: "React" }) &&
+    ((!callee.computed &&
+      t.isIdentifier(callee.property, { name: "createContext" })) ||
+      t.isStringLiteral(callee.property, { value: "createContext" }))
+  );
+}
+
+function buildDisplayNameAssignment(ref, displayName) {
+  return t.assignmentExpression(
+    "=",
+    t.memberExpression(t.cloneNode(ref), t.identifier("displayName")),
+    t.stringLiteral(displayName),
+  );
+}
+
+function addDisplayNameAfterCreateContext(
+  id,
+  path: t.NodePath<t.CallExpression>,
+) {
+  const { parentPath } = path;
+  if (parentPath.isVariableDeclarator()) {
+    // FooContext = React.createContext()
+    const ref = parentPath.node.id;
+    // parentPath.parentPath must be a VariableDeclaration because getDisplayNameReferenceIdentifier
+    // does not support patterns
+    parentPath.parentPath.insertAfter(buildDisplayNameAssignment(ref, id));
+  } else if (parentPath.isAssignmentExpression()) {
+    // var FooContext = React.createContext()
+    const ref = parentPath.node.left;
+    parentPath.insertAfter(buildDisplayNameAssignment(ref, id));
```

Comment:
Inserting after an assignment expression does not preserve completion records. Maybe we should only insert for variable declarations.

Reply:
Inserting after a variable declaration changes the completion record too.

This is a general problem with `insertAfter` and it's not just in this PR. I think we should check it separately checking in `insertAfter` if the completion record matters (e.g. if a parent is a `do` expression) and injeting a temp variable similarly to what we do for `insertAfter` in expressions.

Reply:
The current version of do expression forbids a do block ending with variable declarations: https://tc39.es/proposal-do-expressions/#sec-doexpression-static-semantics-early-errors, but we haven't implemented that in Babel parser. Guess it should be okay for variable declarations (most common cases) here?

Reply:
Uh well ok then, I guess assignment expressions in this case are not common.